atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImpl.java

```diff
@@ -108,25 +111,29 @@ public void reset() {
         return Optional.ofNullable(snapshotMap.get(sequence));
     }
 
-    private void retentionSnapshots() {
-        int snapshotMapSize = snapshotMap.size();
-        if (snapshotMapSize > minimumSize) {
-            int numToRetention = snapshotMapSize - minimumSize;
-            Sequence firstLiveSequence = Optional.ofNullable(Iterables.getFirst(liveSequences.keySet(), null))
-                    .orElse(MAX_VERSION);
+    @VisibleForTesting
+    void retentionSnapshots() {
+        Set<Sequence> currentLiveSequences = liveSequences.keySet();
+        int nonEssentialSnapshotCount = snapshotMap.size() - currentLiveSequences.size();
+        if (nonEssentialSnapshotCount > minimumSize) {
+            int numToRetention = nonEssentialSnapshotCount - minimumSize;
```

Comment:
This is correct I think, in the sense that we want to keep "the live things plus a short tail"... though we probably want to rename `minimumSize` as it isn't frequently reached anymore. Maybe `minimumUnusedSnapshots`?